### DISCUSSION TITLE
The need for a query engine

### DISCUSSION THREAD

**Discussion by tslocke** [rocicorp/replicache#1023] [General]
The shift, in web development, to local-first data, zero-latency UI and realtime multiplayer, seems pretty much inevitable at this point. This is especially true given that, with the right supporting ideas such as those in Replicache, you actually end up with an *easier* programming model compared to the old client/server model of the typical Rails/PHP app.

But one piece of the stack that was central to the old way might be lost - the database's query engine. Sometimes we wrote raw SQL, sometimes via an abstraction such as Rails' ActiveRecord, but we *always* wrote queries. 

As far as I understand, part of the argument for this being OK might be that everything is in memory. We can get the same results with plain old JavaScript: map, filter, helper functions...

But there's two distinct downsides compared to a proper query engine. One is expressivity - a good query language gives you tons of power to express complex rules and constraints succinctly. The other is indexing: the ability to just declare what I'd like indexed, and have the query engine figure out how to efficiently use those indexes to optimise my queries, is another huge productivity boost.

I'm aware of two serious ideas to solve this problem, and bring powerful queries to the local-first/zero-latency/multiplayer party.

One is Instant (https://www.instantdb.com) which aims at a similar value-prop to Replicache but will also, they say, bring datalog to javascript. Datalog in the browser already exists in fact, in DataScript, but ClojureScript remains very niche.

The other is more of a general idea that I saw crop up in a HN thread the other day. Supposedly extensions to the browser filesystem API are coming that will allow sufficiently high performance read/write that compile-to-web-assembly "real" databases will be coming. i.e. SQLite or even Postgres running inside the browser sandbox. As I understand this is all possible already but without the improved file APIs the results are not really practical.

So I'm wondering what the Replicache story is going to be here. Given that Replicache wants to own the low-level client side storage, it would seem it will have to get involved in the query business.

There's one more idea here which I think is going to become extremely important. This is the idea of the "vast key-space" that the app "windows" in to. (See https://roamresearch.com/#/app/aboodman/page/X3mhH5OCs from @aboodman). Obviously, if you're building Twitter, you're not going to sync the entire thing to everyone's browser, and there are no obvious boundaries to chunk things up into spaces. This can all be solved in Replicache because of BYOB, but I'm sure in the future this will become the responsibility of the platform, not something I have to implement myself. Having queries be part of the platform seems to me the right way to do this. If I can execute a single query and get back results from my local cache, if they are there, and from the server if they are not, and if the platform then understands that the results of that query are now part of my window, that would be something.

This was the subject of an essay that got a lot of interest in the Clojure community, by the DataScript guy:

https://tonsky.me/blog/the-web-after-tomorrow/

Someone is going to take a stab at something like this soon enough. The Instant guys seem well placed to go this way and I'd bet they have plans.

With Reflect, Replicache is taking a step towards being the "whole data platform" as opposed to just a client library. How might this develop?

**Comment by aboodman** [comment]
Yeah. Not gonna lie:

>  If I can execute a single query and get back results from my local cache, if they are there, and from the server if they are not, and if the platform then understands that the results of that query are now part of my window, that would be something.

I get that this is compelling. It's outside of our current vision and I don't know if/how there is a future there for us. It's hard to get all these features in the same box. Buttery smooth multiplayer just wants different architecture than a distributed database, for example. I would really like to find ways to draw lines around smaller primitives and let people compose things, rather than having to do/be everything.

For example it's perfectly reasonable to write an old fashioned query on top of a Replicache backend. Postgres can index JSON. You can marry this to local queries quite readily, just requires some glue. Especially if you only need this for a few cases, it's easy enough.

Getting to your question, I think there are three parts to it:

1. How should I query against Replicache data
2. How should I query against the server
3. Is there a way to merge these abstractions

2 and 3 are easy: (2): the way you do today. (3) sure but you have to do it manually. Perhaps patterns will emerge over time and those patterns might turn into features, but we don't have enough experience yet to say.

===

For (1) there is no reason we can't build a Firebase-style query API on top of Replicache. We have considered this. I think we will probably do it eventually. But a lot of developers are comfortable and like the "state management" model instead, like Jotai, Zustand, etc. This covers many of the same use cases and is less work for us to implement.

@ingar is working on adapting Repliear to use Jotai to demonstrate the pattern, then we will look at it and decide if the dx is there or if we want to do anything more first-class in Replicache right now.

**Reply by aboodman** [reply]
> there is no reason we can't build a Firebase-style query API on top of Replicache. We have considered this. I think we will probably do it eventually.

For clarity, a firebase style query API is kind of surprisingly little work once you have the indexes in place. That is actually the harder part.

SQL is more work because as you say there's all the guessing which indexes to use, and also the parsing. I'm not sure SQL is even desired though.

**Reply by aboodman** [reply]
> For example it's perfectly reasonable to write an old fashioned query on top of a Replicache backend. 

Sorry you just made me realize: maybe there's a way to say something like "I would like the result of `<query>` to be in the sync window". That would be neat.

**Reply by tslocke** [reply]
Yes that sure would be. But to support that for general queries, do you have to re-run all of these queries on every server update, to see if new rows now match? Materialze looks like a very impressive solution to this problem - https://materialize.com/

**Reply by tslocke** [reply]
> a lot of developers are comfortable and like the "state management" model instead

Can you expand a bit on how state management and querying overlap? They seem like different problems to me.

**Reply by aboodman** [reply]
> But to support that for general queries, do you have to re-run all of these queries on every server update, to see if new rows now match?

Naively you have to re-run the query when the inputs change. But for many queries this can end up being very often. Materialize and dataflow in general seem like good paths to improve this.

**Reply by aboodman** [reply]
> Can you expand a bit on how state management and querying overlap? They seem like different problems to me.

Think about Repliear: https://blog.replicache.dev/blog/how-we-built-repliear-compound-filters.

One way to think about what the UI is doing there is that it is issuing reactive queries to some datastore. Another way to think about the same problem is that it is reactively maintaining some computed state.

Systems like jotai, mobx, zustand (, and Solid, right?) allow a developer to express things like:

> I would like to reactively render all the issues that are assigned-to-aaron, in-progress, and high-priority

... by feeding all changes to the datastore through a dependency graph of computations.

Without persistent indexes, it's not as fast for the first query as a "real database". And it might feel like worse dx than a query language depending on your pov.

But Replicache does have persistent indexes, so it can be just as efficient. (Actually this kind of reactive dependency graph is exactly what Materialize does inside, I think).

**Reply by aboodman** [reply]
And if were to write our own little query system, this would be how it would work too.
